Add multi-year domain purchases

umurkontaci opened the pull request:
This PR enables users to purchase domains for more than 1 year. As this is a big change, we're adding a feature flag to enable it when we're ready.


### Screenshot

![image](https://cloud.githubusercontent.com/assets/991022/9464907/5f410020-4b30-11e5-9255-052b8c989a84.png)


### Testing

You'll need to test this with D341-code.

- Add a domain to the cart
- Go to checkout
- Change the year from the select box
- Buy the domain

- Make sure the domain is registered for the duration you have picked
- Make sure you are charged the correct amount
- If you opted for a privacy protection, make sure you are charged for it correctly as well.

- Cancel the domain you've just registered

dmsnell commented:
@umurkontaci looks like we have some merge conflicts - does this need a rebase?

umurkontaci commented:
@dmsnell Yup, thanks for the ping. Rebased now

dmsnell commented on client/lib/cart-values/cart-items.js:
```diff
@@ -319,12 +319,14 @@ function businessPlan( slug, properties ) {
  *
  * @param {Object} productSlug - the unique string that identifies the product
  * @param {string} domain - domain name
+ * @param {Number} volume
```
It seems more natural to me to use the word `quantity` to specify how many of a specific product item we want in the cart. Maybe we can have some others chime in. `volume` sounds a bit odd to me and maybe misleading (though maybe I don't understand this well enough!)

umurkontaci replied:
`volume` is the name that's being used in the WP codebase, hence the usage.

dmsnell commented on client/lib/cart-values/cart-items.js:
```diff
@@ -605,6 +607,10 @@ function getDomainRegistrationsWithoutPrivacy( cart ) {
+function isPrivacyProduct( cartItem ) {
+	return cartItem.product_slug === 'private_whois';
+}
+
```
Just for fun, here's another way to write it that I like for its brevity

```js
const isPrivacyProduct = item => 'private_whois' === item.product_slug;
```

umurkontaci replied:
The rest of the functions in the file have the same `function` syntax. I'd rather keep them in the same shape.

dmsnell commented on client/lib/cart-values/cart-items.js:
```diff
@@ -615,7 +621,7 @@ function getDomainRegistrationsWithoutPrivacy( cart ) {
 function changePrivacyForDomains( cart, domainItems, changeFunction ) {
 	return flow.apply( null, domainItems.map( function( item ) {
-		return changeFunction( domainPrivacyProtection( { domain: item.meta } ) );
+		return changeFunction( domainPrivacyProtection( { domain: item.meta, volume: item.volume } ) );
```
Might be a decent time to add some destructuring and spread to cut down on the typing

```js
return flow( ...domainItems.map(
	( { meta: domain, volume } ) => changeFunction( domainPrivacyProtection( { domain, volume } ) )
) );
```

Actually, we might have an issue with the spread operator, in which case if that's true you should stick with `apply`

dmsnell commented on client/lib/cart-values/cart-items.js:
```diff
@@ -647,6 +653,22 @@ function getIncludedDomain( cartItem ) {
+	return function( cart ) {
+		return React.addons.update( cart, { products: { $apply: setItemVolume } } );
+	};
+}
```
since we are already defining some helpers inside this function, why not add a few more?

```js
function setVolume( item, volume ) {
	const matchesItem = i => (
		i.meta === item.meta &&
		( i.product_id === item.product_id || isPrivacyProduce( item ) )
	);

	const extendIfItemMatches = i => ! matchesItem( i ) ? i : { ...item, volume };
	
	return cart => { ...cart, products: cart.products.map( extendIfItemMatches ) };
}
```

also in this code I didn't use the React immutability helper because it didn't seem to add much here other than make it more confusing for me to read. we now have some great operators in JavaScript that do just what we want for that. It would actually still be more useful if `products` were some deeply-nested property inside of `cart` but that just isn't the case here.

umurkontaci replied:
:+1:

umurkontaci replied:
That won't work because spread operator only works with Arrays, iterables in ES6. JSX uses it but it's limited to ES6. We might expect to use the operator over objects in ES7.

dmsnell commented on client/lib/cart-values/cart-items.js:
```diff
@@ -662,6 +684,7 @@ module.exports = {
 	getDomainMappings,
 	getDomainRegistrations,
 	getDomainRegistrationTld,
+	getDependentProducts: getDependentProducts,
```
there's no need to repeat ourselves here, especially since the other items don't

umurkontaci replied:
:+1:

dmsnell commented on client/lib/cart-values/cart-items.js:
```diff
@@ -683,6 +706,8 @@ module.exports = {
 	hasProduct,
 	hasRenewalItem,
+	setVolume: setVolume,
+	isPrivacyProduct: isPrivacyProduct,
```
see my previous comment about repeating here

dmsnell commented on client/my-sites/upgrades/cart/cart-item.jsx:
```diff
@@ -1,29 +1,39 @@
 	removeFromCart: function( event ) {
 		event.preventDefault();
-		analytics.ga.recordEvent( 'Upgrades', 'Clicked Remove From Cart Icon', 'Product ID', this.props.cartItem.product_id );
+		this.recordEvent( 'remove', this.props.cartItem.product_id );
```
Why are we putting this logging code directly in with the logic when we have a valid Flux system in place? I think we should be able to create a separate "logging" store to watch these actions flow by and record the appropriate events when they occur. If we _do_ do this, then that would make this code much more focussed and without the noise of analytics

umurkontaci replied:
I understand where you're coming from, but that's the pattern we've been using so far in everywhere, I don't think this PR is the right place to solve it.

dmsnell commented on client/my-sites/upgrades/cart/cart-item.jsx:
```diff
@@ -85,6 +95,41 @@ module.exports = React.createClass( {
 
+	handleDomainVolumeSelection: function( event ) {
+		event.preventDefault();
+		let volume = parseInt( event.target.value );
```
it looks like we can use `const` here since `volume` isn't assigned anything after the initial value

dmsnell replied:
also pretty important here is that we should always include our radix when using `parseInt` (at least according to Mozilla and others)

```js
const volume = parseInt( event.target.value, 10 );
```

umurkontaci replied:
`parseInt` uses a different base only when the string starts with a "0". So technically, we're safe here as we're controlling the supplied value. I don't mind adding the radix though.

dmsnell replied:
fair point, but yeah, why not add it? it certainly won't slow us down :wink:

umurkontaci replied:
:+1:

dmsnell commented on client/my-sites/upgrades/cart/cart-item.jsx:
```diff
@@ -85,6 +95,41 @@ module.exports = React.createClass( {
+	},
+
+	getVolumeOptions: function() {
+		return range( 1, 6 ).map( number =>
```
This is just to make sure you want five numbers `1, 2, 3, 4, 5` printed here. It looks like you know what you're doing, but it's a bit sneaky here the way `range()` goes up-to-but-not-over the end value.

A comment would greatly help here, or since the list is so small it could be inlined (though I still prefer to use `range()` for its clarity)

```js
return [1, 2, 3, 4, 5].map( number => … );

return range( 1, 5 + 1 ).map( number => … );

// present 1-year to 5-year subscriptions
return range( 1, 6 ).map( number => … );
```

umurkontaci replied:
:+1:

dmsnell commented on client/my-sites/upgrades/cart/cart-item.jsx:
```diff
@@ -85,6 +95,41 @@ module.exports = React.createClass( {
+			<select name="product-domain-volume"
+							className="product-domain-volume"
+							onChange={ this.handleDomainVolumeSelection }
+							value={ this.props.cartItem.volume }>
```
I think something might be messed up here with the formatting

umurkontaci replied:
:+1: I use 2 space tabs to use space efficiently, Webstorm wanted to align for that. :+1:

dmsnell commented on client/my-sites/upgrades/cart/cart-item.jsx:
```diff
@@ -99,7 +144,8 @@ module.exports = React.createClass( {
 					</span>
-					{ this.removeButton() }
+					{ this.domainVolumeSelection() }
+					{ isCredits( this.props.cartItem ) ? null : this.removeButton() }
```
we should be able to go without the ternary-and-null combination

```js
{ ! isCredits( this.props.cartItem ) && this.removeButton() }
```

dmsnell commented on client/my-sites/upgrades/cart/cart-item.jsx:
```diff
@@ -99,7 +144,8 @@ module.exports = React.createClass( {
 						{ this.price() }
 					</span>
-					{ this.removeButton() }
+					{ this.domainVolumeSelection() }
```
I think it would be nice to inline this here with some helper variables at the top of `render()`

```js
render() {
	const { cartItem } = this.props;
	const volumeSelectionIsEnabled = config.isEnabled( 'upgrades/cart/multi-year-domain' );
	const showVolumeSelection =
		volumeSelectionIsEnabled &&
		( this.isBundlePlanApplied() || cartItem.free_trial || isDomainRegistering( cartItem ) );

	…

	</span>
	{ showVolumeSelection &&
		<select name="product-domain-volume"
			className="product-domain-volume"
			…>
			…
		</select>
	}
}
```

umurkontaci replied:
:+1:

dmsnell commented on client/my-sites/upgrades/cart/cart-item.jsx:
```diff
@@ -113,17 +159,27 @@ module.exports = React.createClass( {
 		} else if ( cartItem.volume === 1 ) {
 			switch ( cartItem.product_slug ) {
 				case 'gapps':
+					hasDomainSuffix = cartItem.product_name.indexOf( cartItem.meta ) > -1;
```
`includes` might be really helpful here (`String.prototype.includes()` as of ES6) if these are string values, **lodash** has a more generic `includes()` if not

```js
hasDomainSuffix = cartItem.product_name.includes( cartItem.meta );
```

umurkontaci replied:
:+1:

dmsnell commented on client/my-sites/upgrades/cart/cart-item.jsx:
```diff
@@ -113,17 +159,27 @@ module.exports = React.createClass( {
 			switch ( cartItem.product_slug ) {
 				case 'gapps':
+					hasDomainSuffix = cartItem.product_name.indexOf( cartItem.meta ) > -1;
+					gappsProductName = hasDomainSuffix ? cartItem.product_name : this.translate( '%(productName)s for %(domain)s', {
```
Could we move the string literal to the new line? It would be easier to read here

umurkontaci replied:
:+1:

dmsnell commented:
thanks for rebasing @umurkontaci - it looks like there are now a few anachronisms after the update and I pointed out one or two, but we should probably make sure the style matches the surrounding code if the old stuff had been updated since this PR was born.

klimeryk commented on config/development.json:
```diff
@@ -93,8 +93,9 @@
+		"upgrades/cart/multi-year-domain": true,
 
-		"manage/customize": true,
+	  "manage/customize": true,
```
You've used spaces here - it will look weird for anyone that has configured `tab` to be expanded to something else than 2 spaces :)

umurkontaci commented:
@dmsnell @klimeryk I've updated and rebased the code. Do you mind taking a look at it?

We're shelving the option to buy multi-year domains for now. But I would rather keep this code merged rather than having to keep track of everything to rebase.

dmsnell commented:
@umurkontaci I'm a bit confused - why would we be merging this code if we aren't enabling the feature? Wouldn't it make more sense to leave the code out of the repository until we need it?

umurkontaci commented:
No it wouldn't. If we are not merging this code, by the time we are ready to launch everything in this code would be irrelevant / stale and I have to do everything all over again.

dmsnell commented:
> If we are not merging this code, by the time we are ready to launch, everything in this code would be irrelevant / stale and I have rewrite the whole thing all over again.

That's kind of my point. If we merge this in, it seems like it will sit as dead code if and until we decide to allow multi-year purchases, and by the time that happens, this will all be irrelevant / stale and we will have to redo it all anyway, so why not just let this branch go (or bookmark it) and defer the development until we need it - otherwise we'll just be adding an unnecessary maintenance burden, unless I misunderstand you

umurkontaci commented:
It will still be less work to do though. We would have to test the stuff to be working again and make small adjustments, but this piece will be included in code refactoring processes. Refactoring a bunch of functions is easier than coming with an old code and adjusting the whole to the 10 times refactored code.

It would be unnecessary maintenance burden if we were adding something we didn't know we'd need it or not. 

This PR has been thoroughly tested in here and in pre-oss. I don't want to throw everyone's work away.

dmsnell commented:
@umurkontaci no hard feeling if you end up merging this in, but if I understand you correctly I can't vote to do so - seems like a bad idea to me. I understand not wanting to throw everyone's work away, but that's not a compelling argument to me to put in dead code and I don't think that's the right way to look at this branch. in any case, this code will live on in `git`

feel free to ping other people for review. I appreciate your work on this.

stephanethomas commented:
I agree with @dmsnell, we shouldn't merge this pull request. This will add extra maintenance and introduce additional complexity for something that is not used - without mentioning that it will mislead developers who don't know this part of the code and who will wonder what it does and how it fits in the whole application.
